### PR TITLE
Add thermal deformation example

### PR DESCRIPTION
**Pull request by streeve**
Adds temperature dependent stretch with new example and general way to include thermal physics

Initially only implemented for PMB and without fracture

- [x] Temperature communication
- [x] Thermal strain energy 

~Blocked by #89~

### CODE REVIEW

File: src/CabanaPD_Particles.hpp [hunk 1]
```diff
@@ -345,8 +347,8 @@ class Particles<MemorySpace, PMB, Dimension>
     {
         return Cabana::slice<0>( _aosoa_vol, "volume" );
     }
-    auto sliceType() { return Cabana::slice<0>( _aosoa_other, "type" ); }
-    auto sliceType() const { return Cabana::slice<0>( _aosoa_other, "type" ); }
+    auto sliceType() { return Cabana::slice<5>( _aosoa_other, "type" ); }
```
[the remark this answers is not in the document]
Comment (streeve) [reply src/CabanaPD_Particles.hpp:350]: See line 105

**Review by pabloseleson — APPROVED** [review]

**Comment by pabloseleson** [comment]
@streeve : Put examples inside folders:
mechanics
thermo-mechanics

File: README.md [hunk 2]
```diff
@@ -129,7 +134,14 @@ run with:
 ./CabanaPD/build/install/bin/CrackBranching
 ```
 
-New examples can be created by using any of the current cases as a template. All inputs are currently specified in the example source files themselves.
+### Thermomechanics
+Examples which demonstrate temperature-dependent mechanics and fracture are with `examples/thermomechanics`.
+
+The first example demonstrates a thermoelastic problem without fracture:
```
Comment (streeve) [inline README.md:140]: @pabloseleson can you add the reference here?

File: src/force/CabanaPD_Force_PMB.hpp [hunk 3]
```diff
@@ -69,17 +69,19 @@
 
 namespace CabanaPD
 {
-template <class ExecutionSpace>
-class Force<ExecutionSpace, ForceModel<PMB, Elastic>>
+template <class ExecutionSpace, class... ModelParams>
```
[the remark this answers is not in the document]
Comment (streeve) [reply src/force/CabanaPD_Force_PMB.hpp:72]: Variadic inputs

File: src/force/CabanaPD_Force_PMB.hpp [hunk 4]
```diff
@@ -103,7 +105,7 @@ class Force<ExecutionSpace, ForceModel<PMB, Elastic>>
                            const NeighListType& neigh_list, const int n_local,
                            ParallelType& neigh_op_tag ) const
     {
-        auto c = _model.c;
+        auto model = _model;
```
[the remark this answers is not in the document]
Comment (streeve) [reply src/force/CabanaPD_Force_PMB.hpp:108]: Gets a copy of the full model (instead of just one value)

**Comment by pabloseleson** [comment]
There is an issue when temp0 is not 0.

**Comment by pabloseleson** [comment]
I tried adding the following. Files were created but with no content:

    // ====================================================
    //                      Outputs
    // ====================================================
    // Output displacement along the x-axis
    createDisplacementProfile( MPI_COMM_WORLD, num_cells[0], 0,
                               "xdisplacement_profile.txt", *particles );

    // Output displacement along the y-axis
    createDisplacementProfile( MPI_COMM_WORLD, num_cells[0], 1,
                               "ydisplacement_profile.txt", *particles );

File: src/CabanaPD_ForceModels.hpp [hunk 5]
```diff
@@ -23,9 +25,54 @@ struct BaseForceModel
         : delta( _delta ){};
 
     BaseForceModel( const BaseForceModel& model ) { delta = model.delta; }
+
+    // No-op for temperature.
```
Comment (pabloseleson) [inline src/CabanaPD_ForceModels.hpp:29]: What does it mean "No-op for temperature."?

File: examples/thermomechanics/thermal_deformation.cpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        inputs, particles, force_model );
+
+    // ====================================================
+    //                   Imposed field
+    // ====================================================
+    auto x = particles->sliceReferencePosition();
+    auto temp = particles->sliceTemperature();
+    const double low_corner_y = low_corner[1];
+    // This is purposely delayed until after solver init so that ghosted
+    // particles are correctly taken into account for lambda capture here.
```
Comment (pabloseleson) [inline examples/thermomechanics/thermal_deformation.cpp:101]: "lambda capture" may not be clear for someone not familiar with KOKKOS_LAMBDA.

File: src/CabanaPD_Particles.hpp [hunk 7]
```diff
@@ -646,6 +651,133 @@ class Particles<MemorySpace, LPS, Dimension>
...
+    // Per type.
+    using base_type::n_types;
+
+    // Simulation total domain.
+    using base_type::global_mesh_ext;
+
+    // Simulation sub domain (single MPI rank).
+    using base_type::ghost_mesh_hi;
+    using base_type::ghost_mesh_lo;
+    using base_type::local_mesh_ext;
```
Comment (pabloseleson) [inline src/CabanaPD_Particles.hpp:686]: What is "local_mesh_ext"?

File: src/CabanaPD_Solver.hpp [hunk 8]
```diff
@@ -462,7 +556,7 @@ class SolverFracture
                           neigh_iter_tag{} );
 
             // Add boundary condition.
-            boundary_condition.apply( exec_space{}, *particles );
+            boundary_condition.apply( exec_space{}, *particles, step * dt );
```
Comment (pabloseleson) [inline src/CabanaPD_Solver.hpp:559]: I am wondering why here we do not distinguish between:

if ( !boundary_condition.forceUpdate() )

and 

if ( boundary_condition.forceUpdate() )

like in lines 518-524

File: src/force/CabanaPD_ForceModels_PMB.hpp [hunk 9]
```diff
@@ -121,6 +128,56 @@ struct ForceModel<LinearPMB, Fracture> : public ForceModel<PMB, Fracture>
...
+    using base_temperature_type = BaseTemperatureModel<TemperatureType>;
+    using base_model = PMB;
+    using fracture_type = Elastic;
+    using thermal_type = TemperatureDependent;
+
+    using base_type::c;
+    using base_type::delta;
+    using base_type::K;
+
+    // Thermal coefficients
```
Comment (pabloseleson) [inline src/force/CabanaPD_ForceModels_PMB.hpp:146]: temp0 is not a coefficient. Perhaps say "// Thermal parameters"?

File: examples/thermomechanics/inputs/thermal_deformation.json [hunk 10]
```diff
@@ -4,7 +4,7 @@
     "density"               : {"value": 3980,  "unit": "kg/m^3"},
     "elastic_modulus"       : {"value": 370e+9, "unit": "Pa"},
     "thermal_coefficient"   : {"value": 7.5E-6, "unit": "oC^{-1}"},
-    "reference_temperature" : {"value": 0.0, "unit": "oC"},
+    "reference_temperature" : {"value": 20.0, "unit": "oC"},
```
Comment (streeve) [inline examples/thermomechanics/inputs/thermal_deformation.json:7]: Why 20?

File: src/CabanaPD_ForceModels.hpp (hunk 5 above)
Comment (streeve) [reply src/CabanaPD_ForceModels.hpp:29]: Does nothing

File: src/CabanaPD_Particles.hpp (hunk 7 above)
Comment (streeve) [reply src/CabanaPD_Particles.hpp:686]: Extent - I can rename these to be more clear later

File: examples/thermomechanics/thermal_deformation.cpp (hunk 6 above)
Comment (streeve) [reply examples/thermomechanics/thermal_deformation.cpp:101]: That's fair, but I think it needs to be stated. Add more if you can

File: src/force/CabanaPD_ForceModels_PMB.hpp (hunk 9 above)
Comment (streeve) [reply src/force/CabanaPD_ForceModels_PMB.hpp]: Fixed

File: src/CabanaPD_Solver.hpp (hunk 8 above)
Comment (streeve) [reply src/CabanaPD_Solver.hpp]: Fixed

**Review by streeve — COMMENTED** [review]
@pabloseleson ready for final review

File: examples/thermomechanics/thermal_deformation.cpp (hunk 6 above)
Comment (pabloseleson) [reply examples/thermomechanics/thermal_deformation.cpp:101]: How about saying "taken into account for the lambda functor below".?

**Review by pabloseleson — APPROVED** [review]

**Pull request merged**